File: api/download.py
import json
from http.server import BaseHTTPRequestHandler
import requests

class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        try:
            # 1. قراءة الرابط المرسل من ريأكت
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            data = json.loads(post_data)
            tiktok_url = data.get('url')

            if not tiktok_url:
                raise ValueError("No URL provided")

            # 2. تتبع الرابط وتنظيفه (خلف الكواليس)
            # نستخدم allow_redirects=True للوصول للرابط الأصلي
            headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'}
            response = requests.head(tiktok_url, headers=headers, allow_redirects=True, timeout=10)
            
            # تنظيف الرابط من الزوائد (حذف أي شيء بعد علامة الاستفهام)
            clean_url = response.url.split('?')[0]

            # 3. إرسال الرابط النظيف للأداة الجاهزة
            api_url = f"https://www.tikwm.com/api/?url={clean_url}"
            api_response = requests.get(api_url, headers=headers, timeout=10)
            result_data = api_response.json()

            # 4. إرسال النتيجة النهائية لريأكت
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.send_header('Access-Control-Allow-Origin', '*') # لضمان عملها على الموبايل
            self.end_headers()
            
            if result_data.get('code') == 0:
                final_response = {'download_link': result_data['data']['play']}
            else:
                final_response = {'error': 'فشل الاستخراج من المصدر'}

            self.wfile.write(json.dumps(final_response).encode())

        except Exception as e:
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps({'error': str(e)}).encode())

File: api/download.js
import axios from 'axios';

// دالة لتنظيف الرابط وفك الاختصارات (vt.tiktok)
async function getLongUrl(url) {
  try {
    const response = await axios.get(url, {
      maxRedirects: 10,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36'
      }
    });
    return response.request.res.responseUrl || url;
  } catch {
    return url; // إذا فشل فك الاختصار، نستخدم الرابط الأصلي
  }
}

export default async function handler(req, res) {
  if (req.method !== 'POST') return res.status(405).json({ message: 'POST only' });

  const { url } = req.body;
  if (!url) return res.status(400).json({ success: false, message: 'الرابط مفقود' });

  try {
    // 1. فك الرابط المختصر أولاً لضمان أفضل نتائج
    const cleanUrl = await getLongUrl(url.trim());

    // --- المصدر الأول: TikWM (الأسرع عادةً) ---
    try {
      const res1 = await axios.get(`https://www.tikwm.com/api/?url=${encodeURIComponent(cleanUrl)}`);
      if (res1.data && res1.data.code === 0) {
        return res.status(200).json({
          success: true,
          downloadLink: res1.data.data.play,
          videoInfo: { title: res1.data.data.title || 'فيديو تيك توك' },
          longUrl: cleanUrl
        });
      }
    } catch (e) { console.log("Source 1 Failed"); }

    // --- المصدر الثاني: TiklyDown (موثوق جداً) ---
    try {
      const res2 = await axios.get(`https://api.tiklydown.eu.org/api/download?url=${encodeURIComponent(cleanUrl)}`);
      if (res2.data && res2.data.video) {
        return res.status(200).json({
          success: true,
          downloadLink: res2.data.video.noWatermark || res2.data.video.watermark,
          videoInfo: { title: res2.data.title || 'فيديو تيك توك' },
          longUrl: cleanUrl
        });
      }
    } catch (e) { console.log("Source 2 Failed"); }

    // --- المصدر الثالث: عودة للمصدر الأساسي بتنسيق مختلف ---
    try {
      const res3 = await axios.post('https://api.tikwm.com/api/', new URLSearchParams({ url: cleanUrl }));
      if (res3.data && res3.data.code === 0) {
         return res.status(200).json({
           success: true,
           downloadLink: res3.data.data.play,
           videoInfo: { title: res3.data.data.title },
           longUrl: cleanUrl
         });
      }
    } catch (e) { console.log("Source 3 Failed"); }

    throw new Error('جميع المصادر لم تستجب حالياً');

  } catch (error) {
    return res.status(500).json({
      success: false,
      message: 'فشل استخراج الفيديو. جرب رابطاً آخر أو تأكد أن الحساب عام.'
    });
  }
}
